optional load modes for iCE40

benediktibk opened the pull request:
This makes the two load modes for the iCE40 optional and represents therefore a workaround for #77983.

benediktibk commented:
based upon #80846

josuah commented on drivers/fpga/fpga_ice40.c:
```diff
@@ -589,6 +614,21 @@ static int fpga_ice40_init(const struct device *dev)
 	BUILD_ASSERT(FPGA_ICE40_TRAILING_CLOCKS(inst) >= FPGA_ICE40_TRAILING_CLOCKS_MIN);          \
 	BUILD_ASSERT(FPGA_ICE40_TRAILING_CLOCKS(inst) <= UINT8_MAX);                               \
 	BUILD_ASSERT(FPGA_ICE40_MHZ_DELAY_COUNT(inst) >= 0);                                       \
+	BUILD_ASSERT(ICE40_FPGA_LOAD_MODE_GPIO_AVAILABLE ||                                        \
+		     !DT_INST_PROP(inst, load_mode_bitbang));                                      \
+	BUILD_ASSERT(ICE40_FPGA_LOAD_MODE_SPI_AVAILABLE || DT_INST_PROP(inst, load_mode_bitbang)); \
+	BUILD_ASSERT(!DT_INST_PROP(inst, load_mode_bitbang) ||                                     \
+		     DT_INST_NODE_HAS_PROP(inst, creset_gpios));                                   \
+	BUILD_ASSERT(!DT_INST_PROP(inst, load_mode_bitbang) ||                                     \
+		     DT_INST_NODE_HAS_PROP(inst, cdone_gpios));                                    \
+	BUILD_ASSERT(!DT_INST_PROP(inst, load_mode_bitbang) ||                                     \
+		     DT_INST_NODE_HAS_PROP(inst, clk_gpios));                                      \
+	BUILD_ASSERT(!DT_INST_PROP(inst, load_mode_bitbang) ||                                     \
+		     DT_INST_NODE_HAS_PROP(inst, pico_gpios));                                     \
+	BUILD_ASSERT(!DT_INST_PROP(inst, load_mode_bitbang) ||                                     \
+		     DT_INST_NODE_HAS_PROP(inst, gpios_set_reg));                                  \
+	BUILD_ASSERT(!DT_INST_PROP(inst, load_mode_bitbang) ||                                     \
+		     DT_INST_NODE_HAS_PROP(inst, gpios_clear_reg));                                \
```
How about using [2 different compatibles](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/maxim/ds18b20/ds18b20.c#L359-L369) both targeting this same driver?

That way, it might be possible to move all these build time asserts into devicetree bindings, while also enforcing that only one of them is chosen for every given instance.

josuah reviewed (dismissed):
I really like the direction it is taking, but wonder if it would be interesting to split this in 2 drivers in the same file, using [two different compatibles](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/maxim/ds18b20/ds18b20.c#L359-L369):

```c
#define ICE40_FIELDS_COMMON(inst) \
		.creset = GPIO_DT_SPEC_INST_GET(inst, creset_gpios),               \
		.cdone = GPIO_DT_SPEC_INST_GET(inst, cdone_gpios),                 \
		.mhz_delay_count = FPGA_ICE40_MHZ_DELAY_COUNT(inst),               \
		.config_delay_us = FPGA_ICE40_CONFIG_DELAY_US(inst),               \
		.creset_delay_us = FPGA_ICE40_CRESET_DELAY_US(inst),               \
		.leading_clocks = FPGA_ICE40_LEADING_CLOCKS(inst),                 \
		.trailing_clocks = FPGA_ICE40_TRAILING_CLOCKS(inst),               \


#define ICE40_DEFINE_SPI(inst) \
	static struct fpga_ice40_data fpga_ice40_data_##inst;                      \
                                                                                   \
	static const struct fpga_ice40_config fpga_ice40_config_##inst = {         \
		ICE40_FIELDS_COMMON(inst)                                          \
		.bus = SPI_DT_SPEC_INST_GET(inst, SPI_WORD_SET(8) | SPI_TRANSFER_MSB, 0),\
		.load = fpga_ice40_load_spi};                                      \
                                                                                   \
	DEVICE_DT_INST_DEFINE(inst, fpga_ice40_init, NULL, &fpga_ice40_data_##inst,\
			      &fpga_ice40_config_##inst, POST_KERNEL, CONFIG_FPGA_INIT_PRIORITY,\
			      &fpga_ice40_api);

#define ICE40_DEFINE_GPIO(inst) \
	FPGA_ICE40_PINCTRL_DEFINE(inst);                                           \
                                                                                   \
	static struct fpga_ice40_data fpga_ice40_data_##inst;                      \
                                                                                   \
	static const struct fpga_ice40_config fpga_ice40_config_##inst = {         \
		ICE40_FIELDS_COMMON(inst)                                          \
		.clk = GPIO_DT_SPEC_INST_GET_OR(inst, clk_gpios, {0}),             \
		.pico = GPIO_DT_SPEC_INST_GET_OR(inst, pico_gpios, {0}),           \
		.set = FPGA_ICE40_GPIO_PINS(inst, gpios_set_reg),                  \
		.clear = FPGA_ICE40_GPIO_PINS(inst, gpios_clear_reg),              \
		.load = fpga_ice40_load_gpio,                                      \
		FPGA_ICE40_PINCTRL_CONFIG(inst)};                                  \
                                                                                   \
	DEVICE_DT_INST_DEFINE(inst, fpga_ice40_init, NULL, &fpga_ice40_data_##inst,\
			      &fpga_ice40_config_##inst, POST_KERNEL, CONFIG_FPGA_INIT_PRIORITY,\
			      &fpga_ice40_api);


#define DT_DRV_COMPAT lattice_ice40_spi
DT_INST_FOREACH_STATUS_OKAY(ICE40_DEFINE)
#undef DT_DRV_COMPAT

#define DT_DRV_COMPAT lattice_ice40_gpio
DT_INST_FOREACH_STATUS_OKAY(ICE40_DEFINE)
#undef DT_DRV_COMPAT
```

That is a breaking change, though, and the change you introduced are not: +1 with/without splitting the bindings.

benediktibk commented:
> I really like the direction it is taking, but wonder if it would be interesting to split this in 2 drivers in the same file, using [two different compatibles](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/maxim/ds18b20/ds18b20.c#L359-L369):

Interesting idea, and the breaking change should not be a big issue as the driver is still experimental. @cfriedt Preferences from your side?

benediktibk commented on drivers/fpga/Kconfig.ice40:
```diff
@@ -8,3 +9,23 @@ config ICE40_FPGA
+
+if ICE40_FPGA
+
+config ICE40_FGPA_LOAD_MODE_SPI
+	bool "SPI load mode for the iCE40 fpga driver"
+	default y
+	help
+	  Enable support for the SPI load mode of the iCE40 fpga driver.
+	  This will be possible to use only on more high performance MCUs,
+	  as otherwise the timing constraints of the iCE40 cannot be met.
+
+config ICE40_FGPA_LOAD_MODE_GPIO
+	bool "GPIO bitbang load mode for the iCE40 fpga driver"
+	default n
+	help
+	  Enable support for the GPIO bitbang load mode of the iCE40 fpga
+	  driver. This will ensure that also no low performance MCUs the
+	  timing constraints for the iCE40 are met.
```
I think this is not a good idea, because theoretically it could be necessary to have both options available at the same time. And a Kconfig choice would make this exclusive.

cfriedt commented:
Yeah, until we have better support for deinitializing SPI controllers, I guess anything is better than not having multi-instance.

Let's try two DT compats - why not?

benediktibk commented:
> Yeah, until we have better support for deinitializing SPI controllers, I guess anything is better than not having multi-instance.
> 
> Let's try two DT compats - why not?

Okay, I will do so. But I'm gonna wait until #81212 is merged to avoid a conflict.

benediktibk committed:
drivers: fpga: separate drivers of iCE40 for SPI and GPIO bitbang

Separate the current driver for the FPGA iCE40 into two different ones.
One implements only the SPI load mode, the other one only the GPIO
bitbang mode.

Signed-off-by: Benedikt Schmidt <benedikt.schmidt@embedded-solutions.at>

benediktibk commented:
I've separated the drivers into a SPI and GPIO bitbang implementation, with two different compatibles. I've kept the original binding as it is to avoid necessary changes downstream.

benediktibk commented:
Added a bullet point in the release notes for the change in the binding.

josuah commented on drivers/fpga/fpga_ice40_bitbang.c:
```diff
@@ -0,0 +1,289 @@
+	FPGA_ICE40_ASSERT_RANGE_LEADING_CLOCKS(inst);                                              \
+	FPGA_ICE40_ASSERT_RANGE_TRAILING_CLOCKS(inst);                                             \
+	BUILD_ASSERT(DT_INST_PROP(inst, mhz_delay_count) >= 0);                                    \
+	BUILD_ASSERT(DT_INST_NODE_HAS_PROP(inst, creset_gpios));                                   \
+	BUILD_ASSERT(DT_INST_NODE_HAS_PROP(inst, cdone_gpios));                                    \
+	BUILD_ASSERT(DT_INST_NODE_HAS_PROP(inst, clk_gpios));                                      \
+	BUILD_ASSERT(DT_INST_NODE_HAS_PROP(inst, pico_gpios));                                     \
+	BUILD_ASSERT(DT_INST_NODE_HAS_PROP(inst, gpios_set_reg));                                  \
+	BUILD_ASSERT(DT_INST_NODE_HAS_PROP(inst, gpios_clear_reg));                                \
```
I do not remember seeing a way to specify min/max in devicetree, but `required: True` on the bindings guards against them not being undefined.
No harm in checking twice though.

benediktibk replied:
Those are actually not necessary anymore with a required, I will remove them.

benediktibk replied:
done

josuah commented on dts/bindings/fpga/lattice,ice40-fpga-bitbang.yaml:
```diff
@@ -0,0 +1,48 @@
+include: lattice,ice40-fpga-base.yaml
+
+properties:
+  clk-gpios:
+    type: phandle-array
+    description: |
+      SPI Clock GPIO input on iCE40.
+      Example usage:
+      clk-gpios = <&gpio0 5 GPIO_PUSH_PULL>;
+  pico-gpios:
+    type: phandle-array
+    description: |
+      Peripheral-In Controller-Out GPIO input on iCE40.
+      Example usage:
+      pico-gpios = <&gpio0 7 GPIO_PUSH_PULL>;
+  gpios-set-reg:
+    type: int
+    description: |
+      Register address for setting a GPIO.
+      Example usage:
+      gpios-set-reg = <0x60004008>;
+  gpios-clear-reg:
+    type: int
+    description: |
+      Register address for clearing a GPIO.
+      Example usage:
+      gpios-clear-reg = <0x6000400c>;
+  mhz-delay-count:
+    type: int
+    default: 0
+    description: |
+      in order to create a 1 MHz square wave in the following
+      process.
+      while(true) {
+        *gpios_set_reg |= BIT(n);
+        for(int i = mhz_delay_count; i > 0; --i);
+        *gpios_clear_reg |= BIT(n);
+        for(int i = mhz_delay_count; i > 0; --i);
+      }
+      The default disables the delay.
```
It might be possible to put a `required: True` to each of the mandatory fields, and remove them from the `BUILD_ASSERT()`, as a way to catch the errors earlier. I do not see bugs in doing it this way though!

benediktibk replied:
done

josuah reviewed (dismissed):
Thank you very much for this effort. I could as well be doing it since I suggested the changes...

The situation seems very clear now. Alot of `#ifdef` and `BUILD_ASSERT()` removed.
